geode-for-redis/src/main/java/org/apache/geode/redis/internal/data/RedisSet.java

```diff
@@ -161,49 +162,38 @@ public RedisSet() {}
     return popped;
   }
 
-  public Collection<byte[]> srandmember(int count) {
-    int membersSize = members.size();
+  public List<byte[]> srandmember(int count) {
     boolean duplicatesAllowed = count < 0;
-    if (duplicatesAllowed) {
-      count = -count;
-    }
-
-    if (!duplicatesAllowed && membersSize <= count && count != 1) {
-      return new ArrayList<>(members);
-    }
+    count = duplicatesAllowed ? -count : count;
+    int membersSize = members.size();
+    int memberMapSize = members.getMemberMapSize();
 
     Random rand = new Random();
-
-    // TODO: this could be optimized to take advantage of MemberSet
-    // storing its data in an array. We probably don't need to copy it
-    // into another array here.
-    byte[][] entries = members.toArray(new byte[membersSize][]);
-
-    if (count == 1) {
-      byte[] randEntry = entries[rand.nextInt(entries.length)];
-      // TODO: Now that the result is no longer serialized this could use singleton.
-      // Note using ArrayList because Collections.singleton has serialization issues.
-      List<byte[]> result = new ArrayList<>(1);
-      result.add(randEntry);
-      return result;
-    }
-    if (duplicatesAllowed) {
-      List<byte[]> result = new ArrayList<>(count);
-      while (count > 0) {
-        result.add(entries[rand.nextInt(entries.length)]);
-        count--;
-      }
-      return result;
+    List<byte[]> result = new ArrayList<>(count);
+    if (!duplicatesAllowed && membersSize <= count) {
+      result.addAll(members);
     } else {
-      Set<byte[]> result = new MemberSet(count);
-      // Note that rand.nextInt can return duplicates when "count" is high
-      // so we need to use a Set to collect the results.
-      while (result.size() < count) {
-        byte[] s = entries[rand.nextInt(entries.length)];
-        result.add(s);
+      Set<Integer> usedIndexes = null;
+      if (!duplicatesAllowed) {
+        usedIndexes = new HashSet<>(count);
+      }
+
+      for (int i = 0; i < count; i++) {
+        int randIndex = rand.nextInt(memberMapSize);
+        byte[] member = members.getKeyAtIndex(randIndex);
+
+        while (member == null || (!duplicatesAllowed && usedIndexes.contains(randIndex))) {
```

Comment:
This loop could end up taking a long time to execute if `members` is large, duplicates are not allowed and `count` is close to `members.size()`, as it requires generating a random number that isn't the same as any of the previously generated numbers. Including the outer for loop, this ends up making the srandmember command execute in O(N^2) (or thereabouts) which is not in line with native Redis' quoted complexity of O(N).

For the no duplicates case, native Redis uses a different implementation where, if count is greater than 1/3 of the size of the member set, they instead create a duplicate of the entire set and then remove members from it at random until it has the correct size.  From t_set.c (SRANDMEMBER_SUB_STRATEGY_MUL is defined as 3):
>    /* CASE 3:
     * The number of elements inside the set is not greater than
     * SRANDMEMBER_SUB_STRATEGY_MUL times the number of requested elements.
     * In this case we create a set from scratch with all the elements, and
     * subtract random elements to reach the requested number of elements.
     *
     * This is done because if the number of requested elements is just
     * a bit less than the number of elements in the set, the natural approach
     * used into CASE 4 is highly inefficient. */

It's possible that there's a better approach than this, but it's an improvement over the current one, at least.